Build all artifacts when not doing cross compile

maxamillion opened the pull request:
Signed-off-by: Adam Miller maxamillion@fedoraproject.org

stevekuznetsov commented on hack/build-go.sh:
```diff
@@ -22,6 +22,7 @@ if [[ -z "$@" ]]; then
 fi
 
 os::build::build_binaries "$@"
+OS_GOFLAGS="${OS_GOFLAGS:-} ${OS_IMAGE_COMPILE_GOFLAGS}" os::build::build_static_binaries "${OS_IMAGE_COMPILE_TARGETS[@]-}" "${OS_SCRATCH_IMAGE_COMPILE_TARGETS[@]-}"
```
I don't see a good argument for changing the `hack/test-go.sh` API. Why?

maxamillion replied:
Without the build of those artifacts, the RPM build fails because those are listed as being present and they would be missing without the cross compile as they are only currently produced by the `hack/bulid-cross.sh` script.

stevekuznetsov replied:
Can we achieve that without impacting the most common `make` target for every single developer?

liggitt replied:
Agree, the default invocation of build-go.sh shouldn't build all of these

maxamillion replied:
yeah, I spoke with Clayton about this. I'm going to make a different target called `build-platform` and a `build-platform.sh` ... sorting out some of the details now

liggitt commented on origin.spec:
```diff
@@ -178,7 +178,11 @@ Obsoletes:        openshift-sdn-ovs < %{package_refector_version}
+%if 0%{?make_redistributable}
 %{os_git_vars} hack/build-cross.sh
+%else
+%{os_git_vars} hack/build-go.sh
```
Make sure this doesn't lose building gssapi-enabled client for Linux by default

maxamillion replied:
Will do

maxamillion commented:
Alright, updated with a `build-platform.sh` target and changed the spec file as well.

I have it building in Fedora Koji as a scratch build here for i686, AMD64, and AArch64.

http://koji.fedoraproject.org/koji/taskinfo?taskID=15723228

I had to disable armv7hl because the golang compiler is segfaulting complaining about some assembly something or other but I don't think that's related to Origin's code.

stevekuznetsov commented on hack/build-platform.sh:
```diff
@@ -0,0 +1,48 @@
+platforms=( "${GOHOSTOS}/${GOARCH}" )
+if [[ -n "${OS_ONLY_BUILD_PLATFORMS-}" ]]; then
+  filtered=()
+  for platform in ${platforms[@]}; do
```
```sh
for platform in "${platforms[@]}"; do
                ^               ^
```

maxamillion replied:
:+1:

stevekuznetsov commented on hack/build-platform.sh:
```diff
@@ -0,0 +1,48 @@
+
+source <(go env)
+platforms=( "${GOHOSTOS}/${GOARCH}" )
+if [[ -n "${OS_ONLY_BUILD_PLATFORMS-}" ]]; then
```
```sh
if [[ -n "${OS_ONLY_BUILD_PLATFORMS:-}" ]]; then
                                   ^
```

stevekuznetsov commented on hack/build-platform.sh:
```diff
@@ -0,0 +1,48 @@
+    fi
+  done
+  platforms=("${filtered[@]}")
+  echo "DEBUGGING: ${filtered[@]}"
```
```sh
echo "DEBUGGING: ${filtered[*]}"
                            ^
```

(is this something we want to keep in the script?)

maxamillion replied:
No, that was left by accident.

stevekuznetsov commented on hack/build-platform.sh:
```diff
@@ -0,0 +1,48 @@
+
+# Build the primary client/server for the host platform
+OS_BUILD_PLATFORMS=("${platforms[@]}")
+if [[ "${GOHOSTOS}/${GOARCH}" == "linux/amd64" ]]; then
```
use a switch/case?

maxamillion replied:
Yeah, sorry ... I code primarily in python so long-strings of if statements is just a habit. I'll update.

stevekuznetsov commented on hack/build-platform.sh:
```diff
@@ -0,0 +1,48 @@
+fi
+
+# Pass the necessary tags
+OS_GOFLAGS="${OS_GOFLAGS:-} ${OS_IMAGE_COMPILE_GOFLAGS}" os::build::build_static_binaries "${OS_IMAGE_COMPILE_TARGETS[@]-}" "${OS_SCRATCH_IMAGE_COMPILE_TARGETS[@]-}"
```
```sh
... os::build::build_static_binaries "${OS_IMAGE_COMPILE_TARGETS[@]:-}" "${OS_SCRATCH_IMAGE_COMPILE_TARGETS[@]:-}"
                                                                   ^                                          ^
```

maxamillion replied:
I pulled that from `build-cross.sh` and just left it as it was ... is there a specific issue with it? Should it be fixed in `build-cross.sh` also?

stevekuznetsov replied:
It's a nuanced point, I believe we almost always want the `:-` behavior versus the `-` behavior. We should fix it everywhere, but I wouldn't put that in the scope of this PR.

maxamillion commented:
Is there anything still needing revision on this?

maxamillion commented:
Any updates here?

stevekuznetsov commented:
Sorry for not seeing this earlier. I'm afraid that this new script seems to be duplicating a bunch of logic we already have in `hack/build-{go,cross}.sh`. I think I would like it more if we could make this as thin of a addition on top as possible. If we refactored to allow adding build tags for _all_ architectures (without removing the current per-arch envars) we wouldn't need the `switch/case`, either, and this could collapse even more.

maxamillion commented:
@stevekuznetsov I did it this way by recommendation from @smarterclayton in irc. If needed, I'll just carry this patch downstream until there's a better build framework in place that would allow a build option like this without so much duplication.

stevekuznetsov commented:
@smarterclayton what's the argument against refactoring the underlying bits to allow this to sit on top easily?

smarterclayton commented:
There's no argument against, but would be good to understand what has to be
done.

On Tue, Nov 1, 2016 at 8:26 AM, Steve Kuznetsov <notifications@github.com>
wrote:

> @smarterclayton <https://github.com/smarterclayton> what's the argument
> against refactoring the underlying bits to allow this to sit on top easily?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/10997#issuecomment-257596693>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p29ebv-SulofU1sutamj-2asrvqqks5q51o0gaJpZM4KA5xa>
> .
>

tdawson commented:
I found that setting the variable OS_ONLY_BUILD_PLATFORMS to your platform makes hack/build-cross.sh build only the arch that you tell it.  This gets rid of the need for hack/build-platform.sh.  Thus there is only one hack script needed to be maintained.
I have created a pull request with this change in the origin.spec file, as well as patches needed to build aarch64.
#12333

tdawson commented:
Can this be closed?
I believe this have been fixed in this pull request.
https://github.com/openshift/origin/pull/12869

When building on Fedora, or non-x86_64 environments, only the native arch gets built.
